pkg/kds/service/server.go

```diff
@@ -136,25 +181,38 @@ func (g *GlobalKDSServiceServer) streamEnvoyAdminRPC(
 		return status.Error(codes.Internal, "could not store stream connection")
 	}
 	logger.Info("stored stream connection")
-	for {
-		resp, err := recv()
-		if err == io.EOF {
-			logger.Info("stream stopped")
-			return nil
-		}
-		if status.Code(err) == codes.Canceled {
-			logger.Info("stream cancelled")
-			return nil
-		}
-		if err != nil {
-			logger.Error(err, "could not receive a message")
-			return status.Error(codes.Internal, "could not receive a message")
-		}
-		logger.V(1).Info("Envoy Admin RPC response received", "requestId", resp.GetRequestId())
-		if err := rpc.ResponseReceived(clientID, resp); err != nil {
-			logger.Error(err, "could not mark the response as received")
-			return status.Error(codes.InvalidArgument, "could not mark the response as received")
+	streamResult := make(chan error, 2)
+	go func() {
+		for {
+			resp, err := recv()
+			if err == io.EOF {
+				logger.Info("stream stopped")
+				streamResult <- nil
+				return
+			}
+			if status.Code(err) == codes.Canceled {
+				logger.Info("stream cancelled")
+				streamResult <- nil
+				return
+			}
+			if err != nil {
+				logger.Error(err, "could not receive a message")
+				streamResult <- status.Error(codes.Internal, "could not receive a message")
+				return
+			}
+			logger.V(1).Info("Envoy Admin RPC response received", "requestId", resp.GetRequestId())
+			if err := rpc.ResponseReceived(clientID, resp); err != nil {
+				logger.Error(err, "could not mark the response as received")
+				streamResult <- status.Error(codes.InvalidArgument, "could not mark the response as received")
+				return
+			}
 		}
+	}()
+	select {
+	case <-shouldDisconnectStream.Recv():
```

Comment:
it might be good to log here that we are terminating stream because we did not receive the HC.